Make ::pedestal.http/join? option mandatory in production

josefigueroa-nedap opened the pull request:
## Brief

Fixes #15

Makes the `::pedestal.http/join?` option mandatory when starting the service component. Will throw when it is not present when `default-kind` is anything different than `:dev`.

## QA plan

<!-- Please state a reproducible plan to prove this PR works. Attach screenshots, gifs, etc. if needed. Occasionally, sufficient test coverage removes the need for QAing. -->

## Author checklist

<!-- Please, before publicizing your PR, open it as a "WIP PR", and then review it using the following. -->

* [x] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [x] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [x] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [x] Robustness (red paths, failure handling etc)
  * [x] Test coverage
  * [x] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [x] Breaking API changes

## Reviewer checklist

* [ ] I have checked out this branch and reviewed it locally, running it
* [ ] I have QAed the functionality
* [ ] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes

josefigueroa-nedap committed:
Changing the default production setting to avoid blocking the webserver

jwkoelewijn reviewed (changes requested):
This means the key can be removed from the `dev-map` as well, right? As the `dev-map` is merged into this one if I remember correctly.

vemv commented:
fwiw, I'll take a look at what the issue means

vemv commented:
Not sure if I like the change in spirit. Changing a `true` to `false` or viceversa communicate nothing per se, and don't particularly ensure people understand what's going on in prod.

Introducing a mandatory boolean param, adequately speced/doced forces people to understand the thing being configured and its implications.

Also good to avoid breaking API changes as much as poss.

jwkoelewijn commented:
I think I might agree somewhat...

Would like to point out that the difference in `join?` behaviour between production and dev was undocumented upto this point. So in the very least this should be mentioned somewhere in the documentation.

I think that until we resolve this in a manner that will minimize API breakage, we could change our call-sites to explicitly pass the option.

jwkoelewijn commented:
> and don't particularly ensure people understand what's going on in prod.

Our experience today was that this was already the case 😂

vemv reviewed (changes requested):
Gave this some hammocking.

Note that the `start` function is decoupled from Component (we use defn, not defrecord).

I believe the right thing is:

* if `start` is invoked from Component, don't block
  * so that the System can properly start and stop
* if `start` is invoked directly, block
  * generally, invoking a function doesn't magically create a background execution.

A complete solution might be implemented in those terms.

The alternative would be to add a `join?` option that is mandatory **if** [dev?](https://github.com/nedap/components.pedestal/blob/516fcec31d544ec6ef0d257e6878328ea9b91415/src/nedap/components/pedestal/service/component.clj#L28) is false.

jwkoelewijn commented:
> if start is invoked from Component, don't block
so that the System can properly start and stop

This would work

> generally, invoking a function doesn't magically create a background execution.

I agree with the general observation on semantics

I agree on both points on their own, i.e. in isolation. However, I am a bit afraid that combining both will lead to a "magic" function, doing one thing in one context and something else in another context. It adds some complexity and requires the user to know this and keep this difference in the back of his mind.

Therefor, although it would mean an API, breaking change, I tend more towards the simpler mandatory option.

josefigueroa-nedap commented:
This component is intended to work with component systems. If the system requires to stop stuff in a shutdown hook, then they will not have access to the started system.

Also, the discrepancy between development and production behaviour doesn't help to realise this restriction, because during development your system shutdown perfectly and during production it doesn't and it may fail without reporting anything.

A breaking change like this is fine if you end up with better defaults. If the user of this component really wants a blocking server, he will make the change at his own risk/knowledge.

josefigueroa-nedap commented:
I added some documentation in the spec of the `::pedestal.http/join?`.

josefigueroa-nedap committed:
Added spec and documentation to the keyword ::pedestal.http/join?

josefigueroa-nedap committed:
Fixing documentation for ::pedestal.http/join? keyword

jwkoelewijn commented:
After hammocking on this for a little while I think I start to agree with @josefigueroa-nedap on this, because of the following:

Everything in this project seems to point that it is intended to be used within the context of a stuartsierra component system:
- The Readme explicitly talks about it being used with the Reloaded workflow (which is completely based on stuartsierra component system)
- The readme points to https://github.com/grzm/component.pedestal which in it's own respective readme has the heading "Pedestal Service wrapper for Component Systems"

Combine this with that the _current_ default leaves the production system in an invalid state, I would say that changing the default would make sense.

In a general sense I would agree with having an explicit option for the user to make, however, I think that the notion that it _could_ be used in a different context stems from:

> Note that the start function is decoupled from Component (we use defn, not defrecord).

And tbh, In the more general case (considering goals of this project and the aforementioned readmes), I am starting to feel this is an implementation detail, rather than a feature.

vemv commented on src/nedap/components/pedestal/service/component.clj:
```diff
@@ -9,17 +9,26 @@
+  Warning: Blocking the server thread (true value)can interfere with shutdown
+  hooks that would need access to the started component/system.
+  
+  Default to false (non-blocking)."
```
* blocks the **calling** thread
* component/system -> Component System
* Surround `true` with backticks
* Remove `Default to false (non-blocking)`, it duplicates info

vemv reviewed (changes requested):
I'm OK with the simpler option JW describes (as I can see that having a fine-grained behavior can be seen as 'magical'). But, out of good habits and not setting bad precedents, please let's not break APIs in bad ways

* Chaging true -> false is a breaking change of the bad kind
  * Opaque, implicit
  * if https://github.com/nedap/components.pedestal was a transitive dep of some project, then said project could break by upgrading a _different_ dep
    * RH points out this example as a good reason why breaking changes is never a simple thing to do
* Introducing a new mandatory option is a more transparent breaking change
  * It becomes impossible to have this lib upgraded from a consumer and gave have it broken inadvertently.

Extra mandatory options can be slightly painful, but it seems a fair price we pay for mistakes (in this case, my mistake of leaving prod unconfigured). That way we'll make fewer mistakes :)

jwkoelewijn commented:
> But, out of good habits and not setting bad precedents, please let's not break APIs in bad ways

I agree that we should strive to not break APIs! 
I think, however, that in this specific case it is warranted:

> the current default leaves the production system in an invalid state.

I do not think that anyone is actually using the current default, as it just doesn't make sense.

However, seeing that our own client libraries have already explicitly set the option now, and @vemv's  point about it being a more obvious and transparent API break, I do agree that having it as a mandatory option is the way to go.

jwkoelewijn committed:
Made `::pedestal.http/join` mandatory prod option

`nedap.components.pedestal.service.component/start` will throw when
attempted to be called when the `defaults-kind` is anything but `:dev`
and the `pedestal-options` do not contain the `:pedestal.http/join?`
key.

Added tests.

jwkoelewijn commented:
@josefigueroa-nedap I cannot add you as a reviewer to your PR, but seeing all discussion history was in here, i thought it prudent to add my commit to this PR.

jwkoelewijn committed:
Fix/suppress nvd warning on non-used jquery

thumbnail commented on src/nedap/components/pedestal/service/component.clj:
```diff
@@ -18,14 +19,19 @@
 (speced/defn ^::service/component start [{{::router/keys [routes]} ::router/component
                                           ::service/keys           [defaults-kind pedestal-options expand-routes?]
                                           :as                      ^::service/initialized-component this}]
   (let [dev? (= :dev defaults-kind)
+
+        _ (when (and (not dev?)
+                     (not (spec/valid? ::service/pedestal-production-options pedestal-options)))
+            (throw (ex-info "Invalid :pedestal-options for production"
+                            (spec/explain-data ::service/pedestal-production-options pedestal-options))))
+
         ;; routes may be passed directly (via `::router/component`) or indirectly (via `pedestal-options`). Handle that:
```
```suggestion
  (let [dev? (= :dev defaults-kind)]
        (when dev? (check! ::service/pedestal-production-options pedestal-options))
        ;; routes may be passed directly (via `::router/component`) or indirectly (via `pedestal-options`). Handle that:
```

Indentation might be off, also make sure to require `check!`

jwkoelewijn replied:
good suggestion, somehow forgot about `check!`....

jwkoelewijn committed:
Use check\! instead of valid?

vemv reviewed (approved):
lgtm && kudos for the tests :)

Seems good to squash-merge

jwkoelewijn committed:
Make spec relation more clear

Make `::pedestal-production-options` a `spec/merge`, clarifying the
relation between "regular" `::pedestal-options` and
`::pedestal-production-options`
